src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -40,21 +53,69 @@
   // This list is incomplete
   private static final List<String> IRRELEVANT_WORDS =
       new ArrayList<String>(Arrays.asList("the", "is", "for", "in", "of", "so", "to"));
-
+  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // List of all the tags we are searching for
+    List<String> searchTags = 
+        new ArrayList<String>(Arrays.asList(request.getParameterValues("tags")));
+    // Filter to check if the event has any of tags we're searching for
+    Filter tagsFilter =
+        new FilterPredicate("tags", FilterOperator.IN, searchTags);
+    Query query =
+        new Query("Event").setFilter(tagsFilter);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> events = 
+        new ArrayList<Entity>(results.asList(FetchOptions.Builder.withDefaults()));
+
     // get location
     // filter by location and cutoff outside it
-
+    
     // get tags
-    // drop all without first tag
-    // those with most tags in common with search go to top
+    // drop all without first tag?
+    // Sort list by most tags in common with search
+    Collections.sort(events, new Comparator<Entity>() { 
+      public int compare(Entity o1, Entity o2) {
+        int condition = 
+            tagsInCommon((List<String>) o2.getProperty("tags"), searchTags).compareTo(
+            tagsInCommon((List<String>) o1.getProperty("tags"), searchTags));
+        // For development purposes, if two events have the same number of tags
+        // they are sorted by the event names (which in the test cases are integers)
+        if (condition == 0) {
+          return Integer.compare(
+              Integer.parseInt(o1.getProperty("eventName").toString()),
+              Integer.parseInt(o2.getProperty("eventName").toString()));
+        } else return condition;
+      } 
+    });
     // those closest to the user go to the top
+    
+    // Convert events list to json
+    String json = Utility.convertToJson(events);
+    
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
   }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {}
-
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  }
+  
+  /**
+   * Returns a count of the number of tags two lists have in common
+   *
+   * @return Integer count of number of tags in common
+   * @param tagListA List of tags to be compared
+   * @param tagListB List of tags to be compared
+   */
+  public Integer tagsInCommon(List<String> tagListA, List<String> tagListB) {
```

Comment:
Maybe the method name `intersection` is more readable?

Reply:
Got it!